pandas/io/json/_json.py

```diff
@@ -171,6 +171,34 @@ def _write(
 class SeriesWriter(Writer):
     _default_orient = "index"
 
+    def __init__(
+        self,
+        obj,
+        orient,
+        date_format,
+        double_precision,
+        ensure_ascii,
+        date_unit,
+        index,
+        default_handler=None,
+        indent=0,
+    ):
+        super().__init__(
+            obj,
+            orient,
+            date_format,
+            double_precision,
+            ensure_ascii,
+            date_unit,
+            index,
+            default_handler=default_handler,
+            indent=indent,
+        )
+
+        if is_timedelta64_dtype(obj.dtype) and self.date_format == "iso":
+            obj = obj.copy()
+            self.obj = obj.apply(lambda x: x.isoformat())
```

Comment:
Hmm not sure about this approach - this doesn't work if the timedelta is contained within an `object` array right?

Reply:
You're right, it won't. But the modifications in objToJSON.c should take care of that case.

Reply:
Which reminds me, there should probably also be a test for the `astype(object)` case. I will add.

Reply:
Do we need to do this at the Python level is ultimately the question though? This seems like it negates the need to make the change to the extension module since these delta objects would subsequently be strings when iso format is requested and not actually delta objects.

FWIW this is going to be rather non-performant, especially for large DataFrames

Reply:
Well, the python-level change doesn't entirely negate the need for the change in the extension module.

If one does `df.astype(object).to_json(date_format="iso")`, then we do not execute this code in _json.py and instead goes to the modifications made in objToJSON.c.

So in the case where we pass an object array into the extension module, we can take care of the serialization in C.  In the case where we attempt to serialize the DataFrame directly, timedeltas appear differently.  Namely, they come into the C code as NumPy timedeltas for which there is no method to serialize as ISO.  We *could* perhaps do a conversion to a Pandas timedelta in the C code and then use the `isoformat()` method inside the extension module.

I went with this python-level change based on earlier discussions we had about this issue.  In particular: "This would mean dispatching to vectorized methods in Python where applicable (i.e. a DatetimeArray is converted to an array of ISO strings before getting to the serializer)..."

We do end up with a less performant result in the case that we directly serialize the DataFrame but may gain in simplicity.  The fix for the `object` dtype case should still be good performance-wise.

That said, I would be amenable to attempting a fix contained entirely within the extension module.

Reply:
Hmm OK. Yea this is a weird in between. Maybe OK for now knowing it fixes the bug and we can worry about performance in a follow up